presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public final class SortedPositionLinks
+        implements PositionLinks
+{
+    private final PositionLinks positionLinks;
+    private final int[][] compactedPositionLinks;
+    private final Map<Integer, List<Integer>> positionLinksBuilder;
+
+    private final Comparator<Integer> comparator;
+    private JoinFilterFunction filterFunction;
+
+    public SortedPositionLinks(
+            int size,
+            Comparator<Integer> comparator)
+    {
+        this.comparator = comparator;
+        positionLinks = new PositionLinksList(size);
+        compactedPositionLinks = new int[size][];
+        positionLinksBuilder = new HashMap<>();
+    }
+
+    @Override
+    public void compact()
+    {
+        for (Map.Entry<Integer, List<Integer>> entry : positionLinksBuilder.entrySet()) {
+            Integer key = entry.getKey();
+            List<Integer> positions = entry.getValue();
+            Collections.sort(positions, comparator);
+
+            compactedPositionLinks[key] = new int[positions.size()];
+            for (int i = 0; i < positions.size(); i++) {
+                compactedPositionLinks[key][i] = positions.get(i);
+            }
+
+            for (int i = positions.size() - 2; i >= 0; i--) {
+                positionLinks.link(positions.get(i), positions.get(i + 1));
+            }
+        }
+    }
+
+    @Override
+    public void setFilterFunction(Optional<JoinFilterFunction> filterFunction)
+    {
+        checkState(filterFunction.isPresent());
+        this.filterFunction = filterFunction.get();
+    }
+
+    @Override
+    public void link(int from, int to)
+    {
+        List<Integer> links = positionLinksBuilder.containsKey(to) ? positionLinksBuilder.get(to) : new ArrayList<>();
+        links.add(to);
+        positionLinksBuilder.put(from, links);
+        positionLinksBuilder.remove(to);
+    }
+
+    @Override
+    public int next(int position, int probePosition, Page allProbeChannelsPage)
+    {
+        int nextPosition = positionLinks.next(position, probePosition, allProbeChannelsPage);
+        if (nextPosition < 0) {
+            return -1;
+        }
+        if (applyFilterFilterFunction(nextPosition, probePosition, allProbeChannelsPage)) {
+            return nextPosition;
+        }
+        return -1;
+    }
+
+    @Override
+    public int start(int startingPosition, int probePosition, Page allProbeChannelsPage)
```

Comment:
should not you use `from` as you use such name in other methods. Does it mean something?